src/codec/framers/character_delimited.rs

```diff
@@ -0,0 +1,317 @@
+                .iter()
+                .position(|b| *b as char == self.delimiter);
+
+            match (self.is_discarding, newline_pos) {
+                (true, Some(offset)) => {
+                    // If we found a newline, discard up to that offset and
+                    // then stop discarding. On the next iteration, we'll try
+                    // to read a line normally.
+                    buf.advance(offset + self.next_index + 1);
+                    self.is_discarding = false;
+                    self.next_index = 0;
+                }
+                (true, None) => {
+                    // Otherwise, we didn't find a newline, so we'll discard
+                    // everything we read. On the next iteration, we'll continue
+                    // discarding up to max_len bytes unless we find a newline.
+                    buf.advance(read_to);
+                    self.next_index = 0;
+                    if buf.is_empty() {
+                        return Ok(None);
+                    }
+                }
+                (false, Some(pos)) => {
+                    // We found a correct frame
+
+                    let newpos_index = pos + self.next_index;
+                    self.next_index = 0;
+                    let mut frame = buf.split_to(newpos_index + 1);
+
+                    trace!(
+                        message = "Decoding the frame.",
+                        bytes_proccesed = frame.len()
+                    );
+
+                    let frame = frame.split_to(frame.len() - 1);
+
+                    return Ok(Some(frame.freeze()));
+                }
+                (false, None) if buf.len() > self.max_length => {
+                    // We reached the max length without finding the
+                    // delimiter so must discard the rest until we
+                    // reach the next delimiter
+                    self.is_discarding = true;
+                    warn!(
+                        message = "Discarding frame larger than max_length.",
+                        buf_len = buf.len(),
+                        max_length = self.max_length,
+                        internal_log_rate_secs = 30
+                    );
+                    return Ok(None);
+                }
+                (false, None) => {
+                    // We didn't find the delimiter and didn't
+                    // reach the max frame length.
+                    self.next_index = read_to;
+                    return Ok(None);
+                }
+            }
+        }
+    }
+
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Bytes>, Self::Error> {
+        let frame = match self.decode(buf)? {
+            Some(frame) => Some(frame),
+            None if !buf.is_empty() && !self.is_discarding => {
+                let frame = buf.split_to(buf.len());
+                self.next_index = 0;
+
+                Some(frame.into())
+            }
+            _ => None,
+        };
```

Comment:
Can we use adapters here, or does it cause borrow check issues?

```suggestion
        let frame = self.decode(buf)?
            .or_else(|| {
                (!buf.is_empty() && !self.is_discarding).then(|| {
                    self.next_index = 0;
                    buf.split_to(buf.len()).into()
                }
            });
```

Reply:
Changed in 40183c2915280331443ff08fb25e968911d20748